go/proxyd/latestblock.go

```diff
@@ -0,0 +1,61 @@
+package proxyd
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+type LatestBlockHead struct {
+	url  string
+	quit chan struct{}
+
+	mutex sync.Mutex
+	head  *types.Header
+}
+
+func newLatestBlockHead(url string) *LatestBlockHead {
+	return &LatestBlockHead{
+		url:  url,
+		quit: make(chan struct{}),
+	}
+}
+
+func (h *LatestBlockHead) Start() error {
+	client, err := ethclient.DialContext(context.Background(), h.url)
```

Comment:
Perhaps this `ethclient` could be abstracted into a `ChainReader` so that this functionality can be tested with the `SimulatedBackend`
- https://github.com/ethereum/go-ethereum/blob/0169d579d0eed4f6366697985a7b0f0b99402783/interfaces.go#L54
- https://github.com/ethereum/go-ethereum/blob/abc74a5ffeb7e211954178e5d7b8543d5bd3d3cc/accounts/abi/bind/backends/simulated.go#L745

Its really important to have this thoroughly tested because users trust data returned from remote RPCs based on dns authority + tls (no client side validation on the blockchain data itself) and if we return bad data then users will just trust it blindly

Reply:
@tynes I've tweaked the patch to manually poll `eth_blockNumber` via HTTP in response to this [comment](https://github.com/ethereum-optimism/optimism/pull/2002#discussion_r784201458).